File: backend/shared/schemas/scene_dsl.py
from __future__ import annotations

from typing import Annotated, Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

DslIdentifier = Annotated[str, Field(pattern=r"^[A-Za-z_][A-Za-z0-9_]*$", max_length=100)]


class DslModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Position(DslModel):
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0
    relative_to: str | None = None  # e.g., "left_of", "right_of", "above", "below", "to_edge"
    target_id: str | None = None
    buff: float = 0.2


class ThemeConfig(DslModel):
    primary_color: str = "BLUE"
    secondary_color: str = "GREEN"
    background_color: str = "BLACK"
    font: str | None = None


class CameraState(DslModel):
    position: tuple[float, float, float] | None = None
    zoom: float | None = None


class TransitionSpec(DslModel):
    transition_type: str = "fade_out"  # "fade_out", "slide_out"
    duration: float = 1.0


class VisualElement(DslModel):
    id: DslIdentifier
    type: DslIdentifier  # maps to primitives like "get_text_panel", "get_array_block"
    params: dict[str, Any] = Field(default_factory=dict)
    position: Position | None = None


class AnimationStep(DslModel):
    target_ids: list[DslIdentifier]
    animation_type: DslIdentifier
    params: dict[str, Any] = Field(default_factory=dict)
    run_time: float | None = None
    simultaneous: bool = False


class SceneDSLBeat(DslModel):
    id: DslIdentifier
    label: str
    duration_seconds: float = Field(gt=0, le=3600)
    narration: str | None = None
    visual_elements: list[VisualElement] = Field(default_factory=list)
    animations: list[AnimationStep] = Field(default_factory=list)
    camera: CameraState | None = None
    transition_out: TransitionSpec | None = None


class SceneDSL(DslModel):
    version: str = "1.0"
    title: str
    beats: list[SceneDSLBeat] = Field(default_factory=list)
    global_theme: ThemeConfig | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_references(self) -> SceneDSL:
        beat_ids: set[str] = set()
        known_elements: set[str] = set()
        for beat in self.beats:
            if beat.id in beat_ids:
                raise ValueError(f"Duplicate beat id: {beat.id}")
            beat_ids.add(beat.id)
            for element in beat.visual_elements:
                if element.id in known_elements:
                    raise ValueError(f"Duplicate visual element id: {element.id}")
                if element.position and element.position.target_id:
                    if element.position.target_id not in known_elements:
                        raise ValueError(
                            f"Unknown position target_id: {element.position.target_id}"
                        )
                known_elements.add(element.id)
            for animation in beat.animations:
                missing = set(animation.target_ids) - known_elements
                if missing:
                    raise ValueError(f"Unknown animation target ids: {sorted(missing)}")
        return self

File: frontend/src/utils/sceneDsl.ts
import type {
  DslAnimationStep,
  DslCameraState,
  DslPosition,
  DslTheme,
  DslTransition,
  DslVisualElement,
  JsonValue,
  SceneDsl,
  SceneDslBeat,
} from '../types/api';

export const pythonLiteral = (value: JsonValue): string => {
  if (value === null) return 'None';
  if (typeof value === 'boolean') return value ? 'True' : 'False';
  if (typeof value === 'string') return JSON.stringify(value);
  if (typeof value === 'number') return Number.isFinite(value) ? String(value) : '0.0';
  if (Array.isArray(value)) return `[${value.map(pythonLiteral).join(', ')}]`;
  return `{${Object.entries(value)
    .map(([key, item]) => `${JSON.stringify(key)}: ${pythonLiteral(item)}`)
    .join(', ')}}`;
};

const formatPosition = (position: DslPosition | null): string => position
  ? `Position(x=${position.x}, y=${position.y}, z=${position.z}, relative_to=${pythonLiteral(position.relative_to)}, target_id=${pythonLiteral(position.target_id)}, buff=${position.buff})`
  : 'None';

const formatTheme = (theme: DslTheme | null): string => theme
  ? `ThemeConfig(primary_color=${pythonLiteral(theme.primary_color)}, secondary_color=${pythonLiteral(theme.secondary_color)}, background_color=${pythonLiteral(theme.background_color)}, font=${pythonLiteral(theme.font)})`
  : 'None';

const formatCamera = (camera: DslCameraState | null): string => camera
  ? `CameraState(position=${pythonLiteral(camera.position)}, zoom=${pythonLiteral(camera.zoom)})`
  : 'None';

const formatTransition = (transition: DslTransition | null): string => transition
  ? `TransitionSpec(transition_type=${pythonLiteral(transition.transition_type)}, duration=${transition.duration})`
  : 'None';

const formatVisualElement = (element: DslVisualElement): string => `VisualElement(
                    id=${pythonLiteral(element.id)},
                    type=${pythonLiteral(element.type)},
                    params=${pythonLiteral(element.params)},
                    position=${formatPosition(element.position)}
                )`;

const formatAnimation = (animation: DslAnimationStep): string => `AnimationStep(
                    target_ids=${pythonLiteral(animation.target_ids)},
                    animation_type=${pythonLiteral(animation.animation_type)},
                    params=${pythonLiteral(animation.params)},
                    run_time=${pythonLiteral(animation.run_time)},
                    simultaneous=${pythonLiteral(animation.simultaneous)}
                )`;

const formatBeat = (beat: SceneDslBeat): string => `SceneDSLBeat(
            id=${pythonLiteral(beat.id)},
            label=${pythonLiteral(beat.label)},
            duration_seconds=${beat.duration_seconds},
            narration=${pythonLiteral(beat.narration)},
            visual_elements=[
${beat.visual_elements.map(formatVisualElement).map((item) => `                ${item}`).join(',\n')}
            ],
            animations=[
${beat.animations.map(formatAnimation).map((item) => `                ${item}`).join(',\n')}
            ],
            camera=${formatCamera(beat.camera)},
            transition_out=${formatTransition(beat.transition_out)}
        )`;

export const formatDslToPython = (dsl: SceneDsl): string => `from shared.schemas.scene_dsl import SceneDSLBeat, VisualElement, AnimationStep, Position, ThemeConfig, CameraState, TransitionSpec

class GeneratedSceneDSL:
    version = ${pythonLiteral(dsl.version)}
    title = ${pythonLiteral(dsl.title)}
    global_theme = ${formatTheme(dsl.global_theme)}
    metadata = ${pythonLiteral(dsl.metadata)}
    beats = [
${dsl.beats.map(formatBeat).map((item) => `        ${item}`).join(',\n')}
    ]
`;
